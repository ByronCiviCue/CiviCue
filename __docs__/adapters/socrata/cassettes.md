HTTP Cassettes (record/replay)
- Modes: record (capture live responses), replay (serve fixtures), passthrough (real network).
- Helper: `withCassette(testName, fn, { mode })` stores/reads fixtures at `tests/fixtures/cassettes/{testName}.json`.
- Redaction: Authorization header is omitted during recording.
- CI: run in replay mode only to avoid network. Regenerate fixtures locally in record mode.
- Keep fixtures minimal and deterministic (no volatile timestamps).

Usage example
```
await withCassette('v3-client-basic', async () => {
  const res = await v3PostQuery({ domain, datasetId, soql: 'select *' });
  expect(res.rows.length).toBeGreaterThan(0);
}, { mode: 'replay' });
```

