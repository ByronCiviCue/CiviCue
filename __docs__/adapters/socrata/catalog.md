# Socrata Catalog Ingest

This document describes the process for ingesting the global Socrata catalog into the local database.

## Schema

The ingest process populates the following tables:

- `socrata_hosts`: Stores information about Socrata hosts, including their region.
- `socrata_domains`: Stores information about Socrata domains, including their country and region.
- `socrata_agencies`: Stores information about agencies associated with a Socrata host.

The SQL schema can be found in `__docs__/migrations/socrata-catalog.sql`.

## Ingest CLI

The catalog can be ingested using the `bin/socrata-catalog.ts` script.

**Usage:**

```bash
./bin/socrata-catalog.ts [options]
```

**Options:**

- `--regions=US,EU`: Comma-separated list of regions to ingest (default: US,EU).
- `--host=<hostname>`: Filter to a specific host.
- `--limit=<number>`: Maximum number of domains to fetch per region.
- `--dry-run`: Run the ingest process without writing to the database.

**Note:** The `DATABASE_URL` environment variable must be set unless running with `--dry-run`.

**CI Note:** This script is not intended to be run in a CI environment and will exit if `process.env.CI` is set.

## Smoke Test

A smoke test script is available to verify that the client stack can query a dataset.

**Usage:**

```bash
./bin/socrata-smoke.ts --host=<hostname> --dataset=<resource-id> --allow-live
```

**Note:** This script performs live API requests and must be run with the `--allow-live` flag.
## Local registry creation

For local development, you can apply the Socrata catalog SQL to your Postgres instance:

```
node bin/db-apply-sql.mjs --file __docs__/migrations/socrata-catalog.sql
```

Notes:
- Local-only; not used in CI.
- Requires `DATABASE_URL`.
