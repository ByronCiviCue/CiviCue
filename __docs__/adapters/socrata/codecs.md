Contract
- getCodecs(): returns a set of parse/format codecs for common Socrata logical types.
- codecFor(type): returns a single codec by logical type key.
- parse maps raw JSON → canonical runtime; format maps runtime → JSON-safe value.

Notes
- number/money/percent parse finite numbers; invalid inputs return null.
- date/datetime expect ISO-like strings or Date; invalid dates return null.
- checkbox accepts boolean/"true"/"false"/1/0 strings; others return null.
- location/point/polygon accept typical Socrata/GeoJSON shapes, best-effort normalization.

