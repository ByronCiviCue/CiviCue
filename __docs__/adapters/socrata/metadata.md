Contract: Socrata metadata normalization

Exports
- `fetchDatasetMetadata({ domain, datasetId }) => Promise<DatasetMetadata>`
- `normalizeColumn(col)` and `mapLogicalType(apiType, col?)` for reuse

Types
- `LogicalType`: text | number | checkbox | date | datetime | money | percent | url | email | phone | location | point | polygon | json | unknown
- `NormalizedColumn`: { id, name, fieldName, apiType, logicalType, nullable, hidden, description? }
- `DatasetMetadata`: { id, domain, columns: NormalizedColumn[] }

Behavior
- Fetches `https://{domain}/api/views/{datasetId}.json` with `Accept: application/json` and `X-App-Token` when available.
- Parses only documented fields from `columns[]`.
- Logical type mapping is deterministic and does not rely on undocumented fields. Location honors `subColumnType`.
- `nullable` defaults to true; `flags` containing `required` sets to false.

Notes
- No network in tests; use fetch spy.
- Do not edit tasks/config here; adapter-only.

