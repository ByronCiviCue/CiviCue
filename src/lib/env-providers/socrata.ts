/**
 * Resolve Socrata App Token by host with fallback to global.
 * - Exact host override: SOCRATA__{HOST}__APP_TOKEN
 * - Global fallback: SOCRATA_APP_TOKEN
 */
import { resolveSocrataAppToken as _resolve } from '../env.js';

export function resolveSocrataAppToken(host: string): string | undefined {
  return _resolve(host);
}

export function socrataHeadersFor(host: string): Record<string, string> {
  const token = _resolve(host);
  return token ? { 'X-App-Token': token } : {};
}
