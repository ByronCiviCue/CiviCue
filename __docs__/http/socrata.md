# Socrata HTTP Client

Minimal client for interacting with Socrata Open Data API.

## Usage

### Health Check

```typescript
import { healthPing } from '../src/lib/clients/socrata.js';

// Check if a Socrata host is available
const isAvailable = await healthPing('data.sfgov.org');
console.log(`Host available: ${isAvailable}`);
```

### Fetch Dataset Rows

```typescript
import { getRows } from '../src/lib/clients/socrata.js';

// Fetch all rows from a dataset
const allRows = await getRows('data.sfgov.org', 'abc-123');

// Fetch with query options
const filteredRows = await getRows('data.sfgov.org', 'abc-123', {
  limit: 100,
  select: 'id,name,status',
  where: 'status = "APPROVED"',
  order: 'created_date DESC',
  offset: 0
});
```

## Error Handling

Non-2xx responses throw `SocrataHttpError` with status code, URL, and response body snippet:

```typescript
try {
  await getRows('data.sfgov.org', 'invalid-id');
} catch (error) {
  if (error instanceof SocrataHttpError) {
    console.error(`Socrata API error: ${error.status} - ${error.url}`);
  }
}
```

## Authentication

The client automatically includes X-App-Token headers when available via the environment token resolver.