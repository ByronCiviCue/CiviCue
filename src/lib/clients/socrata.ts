import { socrataFetch, buildSocrataUrl, ensureOk } from '../http/socrata.js';

export type GetRowsOptions = {
  limit?: number;
  select?: string;
  where?: string;
  order?: string;
  offset?: number;
};

export async function healthPing(host: string): Promise<boolean> {
  try {
    const url = buildSocrataUrl(host, '/api/catalog/v1', { limit: 1 });
    const response = await socrataFetch(url.toString());
    if (response.status === 404) {
      return false;
    }
    await ensureOk(response);
    return true;
  } catch (error) {
    if (error instanceof Error && error.message.includes('404')) {
      return false;
    }
    throw error;
  }
}

export async function getRows<T = unknown>(
  host: string,
  datasetId: string,
  opts?: GetRowsOptions
): Promise<T[]> {
  const queryParams: Record<string, string | number | undefined> = {};
  
  if (opts?.limit !== undefined) {
    queryParams['$limit'] = opts.limit;
  }
  if (opts?.select !== undefined) {
    queryParams['$select'] = opts.select;
  }
  if (opts?.where !== undefined) {
    queryParams['$where'] = opts.where;
  }
  if (opts?.order !== undefined) {
    queryParams['$order'] = opts.order;
  }
  if (opts?.offset !== undefined) {
    queryParams['$offset'] = opts.offset;
  }

  const url = buildSocrataUrl(host, `/resource/${datasetId}.json`, queryParams);
  const response = await socrataFetch(url.toString());
  const validatedResponse = await ensureOk(response);
  return validatedResponse.json() as Promise<T[]>;
}