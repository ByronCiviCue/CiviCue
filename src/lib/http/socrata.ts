import { socrataHeadersFor } from '../env-providers/socrata.js';

export async function socrataFetch(url: string, init: RequestInit = {}) {
  const u = new URL(url);
  const tokenHeaders = socrataHeadersFor(u.host);
  const baseHeaders: Record<string, string> = { Accept: 'application/json' };
  const merged = { ...baseHeaders, ...tokenHeaders, ...(init.headers as Record<string, string> | undefined) };
  return fetch(u.toString(), { ...init, headers: merged });
}

export function buildSocrataUrl(host: string, path: string, query?: Record<string, string | number | boolean | undefined>): URL {
  const url = new URL(`https://${host}${path}`);
  if (query) {
    Object.entries(query).forEach(([key, value]) => {
      if (value !== undefined) {
        url.searchParams.set(key, String(value));
      }
    });
  }
  return url;
}

export class SocrataHttpError extends Error {
  constructor(
    readonly status: number,
    readonly url: string,
    readonly bodySnippet?: string
  ) {
    super(`Socrata ${status}: ${url}`);
  }
}

async function ensureOk(res: Response): Promise<Response> {
  if (res.ok) return res;
  const text = await res.text().catch(() => '');
  throw new SocrataHttpError(res.status, res.url, text.slice(0, 512));
}

export { ensureOk };

