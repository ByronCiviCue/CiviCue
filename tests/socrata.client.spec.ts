import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { healthPing, getRows } from '../src/lib/clients/socrata.js';
import { SocrataHttpError } from '../src/lib/http/socrata.js';
import * as socrataProviders from '../src/lib/env-providers/socrata.js';

describe('Socrata Client', () => {
  let fetchSpy: any;
  let headersSpy: any;

  beforeEach(() => {
    fetchSpy = vi.spyOn(globalThis, 'fetch');
    headersSpy = vi.spyOn(socrataProviders, 'socrataHeadersFor');
  });

  afterEach(() => {
    fetchSpy.mockRestore();
    headersSpy.mockRestore();
  });

  describe('healthPing', () => {
    it('should build correct URL and return true for successful response', async () => {
      headersSpy.mockReturnValue({ 'X-App-Token': 'test-token' });
      fetchSpy.mockResolvedValue(new Response('{"results":[]}', { status: 200 }));

      const result = await healthPing('data.sfgov.org');

      expect(result).toBe(true);
      expect(fetchSpy).toHaveBeenCalledWith(
        'https://data.sfgov.org/api/catalog/v1?limit=1',
        {
          headers: {
            Accept: 'application/json',
            'X-App-Token': 'test-token'
          }
        }
      );
    });

    it('should return false for 404 responses', async () => {
      headersSpy.mockReturnValue({});
      fetchSpy.mockResolvedValue(new Response('Not Found', { status: 404 }));

      const result = await healthPing('invalid.host');

      expect(result).toBe(false);
    });

    it('should throw for 5xx responses', async () => {
      headersSpy.mockReturnValue({});
      const mockResponse = new Response('Server Error', { status: 500 });
      Object.defineProperty(mockResponse, 'url', { value: 'https://test.com' });
      fetchSpy.mockResolvedValue(mockResponse);

      await expect(healthPing('test.com')).rejects.toThrow(SocrataHttpError);
    });
  });

  describe('getRows', () => {
    it('should build URL with dataset ID and no options', async () => {
      headersSpy.mockReturnValue({});
      fetchSpy.mockResolvedValue(new Response('[{"id":"123"}]', { status: 200 }));

      await getRows('data.sfgov.org', 'abc-123');

      expect(fetchSpy).toHaveBeenCalledWith(
        'https://data.sfgov.org/resource/abc-123.json',
        {
          headers: {
            Accept: 'application/json'
          }
        }
      );
    });

    it('should serialize defined query options to Socrata parameters', async () => {
      headersSpy.mockReturnValue({ 'X-App-Token': 'test-token' });
      fetchSpy.mockResolvedValue(new Response('[{"id":"123"}]', { status: 200 }));

      await getRows('data.sfgov.org', 'abc-123', {
        limit: 100,
        select: 'id,name',
        where: 'status = "APPROVED"',
        order: 'date DESC',
        offset: 50
      });

      const expectedUrl = 'https://data.sfgov.org/resource/abc-123.json?%24limit=100&%24select=id%2Cname&%24where=status+%3D+%22APPROVED%22&%24order=date+DESC&%24offset=50';
      expect(fetchSpy).toHaveBeenCalledWith(
        expectedUrl,
        {
          headers: {
            Accept: 'application/json',
            'X-App-Token': 'test-token'
          }
        }
      );
    });

    it('should omit undefined options from query parameters', async () => {
      headersSpy.mockReturnValue({});
      fetchSpy.mockResolvedValue(new Response('[{"id":"123"}]', { status: 200 }));

      await getRows('data.sfgov.org', 'abc-123', {
        limit: 10,
        select: undefined,
        where: 'active = true'
      });

      expect(fetchSpy).toHaveBeenCalledWith(
        'https://data.sfgov.org/resource/abc-123.json?%24limit=10&%24where=active+%3D+true',
        expect.any(Object)
      );
    });

    it('should return parsed JSON response', async () => {
      const mockData = [{ id: '123', name: 'Test' }];
      headersSpy.mockReturnValue({});
      fetchSpy.mockResolvedValue(new Response(JSON.stringify(mockData), { status: 200 }));

      const result = await getRows('data.sfgov.org', 'abc-123');

      expect(result).toEqual(mockData);
    });

    it('should throw SocrataHttpError for non-2xx responses', async () => {
      headersSpy.mockReturnValue({});
      const mockResponse = new Response('Bad Request', { status: 400 });
      Object.defineProperty(mockResponse, 'url', { value: 'https://test.com/resource/abc.json' });
      fetchSpy.mockResolvedValue(mockResponse);

      await expect(getRows('test.com', 'abc')).rejects.toThrow(SocrataHttpError);
      await expect(getRows('test.com', 'abc')).rejects.toThrow('Socrata 400: https://test.com/resource/abc.json');
    });
  });
});