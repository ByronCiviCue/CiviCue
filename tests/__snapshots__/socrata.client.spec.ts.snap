// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Socrata Client > replays a basic v2 GET page via cassette 1`] = `
[
  {
    "active": true,
    "id": "r1",
    "name": "Alpha",
  },
  {
    "active": false,
    "id": "r2",
    "name": "Beta",
  },
]
`;
